### ISSUE TITLE
I should restart whole DB server when infinite loop

### ISSUE THREAD

**Issue by suricactus** [plv8/plv8#187]
Hello. When I make infinite look like `while(true) {}` I am unable to stop psql until I restart postgres server. Is there workaround to catch signals from like those send by `ctrl-d` and `ctrl-c`?

**Comment by ed-g** [comment]
To reproduce:

```sql
do $$ while (true) {}; $$ language plv8;
```

`pg_terminate_backend(pid)` and `pg_cancel_backend(pid)` have no effect.

**Comment by citizenfish** [comment]
I raised this as a feature request last year.

**Comment by suricactus** [comment]
I think it is a huge issue if there is no way to stop plv8 by OS signal and critical bug.

**Comment by suricactus** [comment]
Do you plan to fix this issue anytime soon?

**Comment by bs85** [comment]
@JerrySievert any feedback on this? Is there a technical reason making this complicated?

**Comment by JerrySievert** [comment]
cross-platform threading and v8 internal changes have made it a little more challenging than I would like.

I could probably *hack* a solution in, but I dislike doing that.

**Comment by bs85** [comment]
I guess I'll just put additional guards at the application level then, thank you for the quick answer!

**Comment by JerrySievert** [comment]
I have something working right now, but as I said, it's a hack.  looking at options for converting it from a hack into something that will work on all platforms reliably.

**Comment by JerrySievert** [comment]
@bs85 et al, if you're up for it, check out 'https://github.com/jerrysievert/plv8/tree/execution_timeout':

```shell
$ git clone https://github.com/JerrySievert/plv8.git
$ cd plv8
$ git checkout execution_timeout
$ make EXECUTION_TIMEOUT=1
```

you can control the execution timeout using the `plv8.execution_timeout` variable, the default is `300` seconds. (it is user configurable from 1 second to 65536 seconds)

I have tried this out in macOS and linux, but have not tried in windows - likely I'll need to implement the threading in a different manner (hence it not being a PR or merged against this repo yet).

give it a try, I'd like some feedback before I look at merging it in upstream.

**Comment by bs85** [comment]
I built and tested this on linux, seems to work as intended. I think this is a really good safeguard in production, and having it available as a variable seems like the right way to do it.

I don't know anything about threading on Windows, but is there anything specific I could test to help?

**Comment by JerrySievert** [comment]
nope, won't compile in windows - I'll likely have to write a windows-specific port of the functionality due to the way that threads are handled.

I'll see what I can do, but found an issue during the port that I'll need to fix before a release.  likely I'll release that as `2.3.1`, with notes for enabling, and make it default for `2.4.0`.

**Comment by BlackYoup** [comment]
@JerrySievert Hello! Are there some blockers for a 2.3.1 release?

**Comment by JerrySievert** [comment]
not really - I was hoping to have a windows version as well, but I guess I can go ahead and release the 2.3.1 release.

**Comment by suricactus** [comment]
I'm not sure if this bug should be closed. as far as I understand, there is still no way to stop plv8 by sending signals?

**Comment by suricactus** [comment]
Using `kill -9` will kill entire postgres backend and only God knows if it would be able to restore.

If I write `do $$ while (true) {}; $$ language plv8;` in psql, there should be some way to stop it by pressing `Ctrl+d`, `Ctrl+d` or `pg_terminate_backend(pid)` and `pg_cancel_backend(pid)`.

**Comment by JerrySievert** [comment]
it will kill that single process that was running that query, yes, not the entire database engine itself, since postgres forks for every connection.

**Comment by bs85** [comment]
I think what he means is that killing the process will cause the entire postgres cluster to restart and go into recovery mode (which happens anytime a backend doesn't terminate properly)

**Comment by JerrySievert** [comment]
compile against master. signals now terminate the v8 subsystem.

**Comment by bs85** [comment]
Amazing, thank you very much!

**Comment by dotjosh** [comment]
Will the signal handlers still be honored if I don't compile with EXECUTION_TIMEOUT?

**Comment by JerrySievert** [comment]
yes - see: https://github.com/plv8/plv8/blob/r3.0alpha/plv8.cc#L524

`EXECUTION_TIMEOUT` itself adds a timer thread that then terminates execution, different code path.
